File: shared/node_libs/src/validation/string.ts
/**
 * String validators for input validation.
 */

import { type ValidationResult, type Validator, success, failure } from './index.js';

export interface NotEmptyOptions {
  errorMessage?: string;
}

/**
 * Validates that a string is not empty or whitespace-only.
 * Returns the trimmed string.
 */
export function notEmpty(options: NotEmptyOptions = {}): Validator<string, string> {
  const errorMessage = options.errorMessage ?? 'Value cannot be empty';

  return (value: string): ValidationResult<string> => {
    if (typeof value !== 'string') {
      return failure('Value must be a string');
    }

    const trimmed = value.trim();
    if (trimmed === '') {
      return failure(errorMessage);
    }

    return success(trimmed);
  };
}

export interface LengthOptions {
  errorMessage?: string;
}

/**
 * Validates that a string length is within a range.
 * @param minLen - Minimum length (inclusive)
 * @param maxLen - Maximum length (inclusive), 0 for unlimited
 */
export function length(minLen: number, maxLen: number = 0, options: LengthOptions = {}): Validator<string, string> {
  return (value: string): ValidationResult<string> => {
    if (typeof value !== 'string') {
      return failure('Value must be a string');
    }

    const len = value.length;

    if (len < minLen) {
      return failure(options.errorMessage ?? `Value must be at least ${minLen} characters`);
    }

    if (maxLen > 0 && len > maxLen) {
      return failure(options.errorMessage ?? `Value must be at most ${maxLen} characters`);
    }

    return success(value);
  };
}

/**
 * Validates minimum string length.
 */
export function minLength(min: number, options: LengthOptions = {}): Validator<string, string> {
  return length(min, 0, options);
}

/**
 * Validates maximum string length.
 */
export function maxLength(max: number, options: LengthOptions = {}): Validator<string, string> {
  return length(0, max, options);
}

export interface MatchOptions {
  errorMessage?: string;
}

/**
 * Validates that a string matches a regex pattern.
 */
export function match(pattern: RegExp | string, options: MatchOptions = {}): Validator<string, string> {
  const regex = typeof pattern === 'string' ? new RegExp(pattern) : pattern;
  const errorMessage = options.errorMessage ?? 'Value does not match required pattern';

  return (value: string): ValidationResult<string> => {
    if (typeof value !== 'string') {
      return failure('Value must be a string');
    }

    if (!regex.test(value)) {
      return failure(errorMessage);
    }

    return success(value);
  };
}

export interface AlphanumericOptions {
  allowUnderscore?: boolean;
  allowDash?: boolean;
  errorMessage?: string;
}

/**
 * Validates that a string contains only alphanumeric characters.
 */
export function alphanumeric(options: AlphanumericOptions = {}): Validator<string, string> {
  const { allowUnderscore = false, allowDash = false } = options;
  const errorMessage = options.errorMessage ?? 'Value must contain only alphanumeric characters';

  let chars = 'a-zA-Z0-9';
  if (allowUnderscore) chars += '_';
  if (allowDash) chars += '-';

  const regex = new RegExp(`^[${chars}]+$`);

  return (value: string): ValidationResult<string> => {
    if (typeof value !== 'string') {
      return failure('Value must be a string');
    }

    if (value === '') {
      return failure('Value cannot be empty');
    }

    if (!regex.test(value)) {
      return failure(errorMessage);
    }

    return success(value);
  };
}

export interface SlugOptions {
  errorMessage?: string;
}

/**
 * Validates that a string is a valid URL slug.
 * A valid slug contains only lowercase letters, numbers, and hyphens.
 */
export function slug(options: SlugOptions = {}): Validator<string, string> {
  const errorMessage = options.errorMessage ?? 'Value must be a valid URL slug';
  const slugRegex = /^[a-z0-9]+(?:-[a-z0-9]+)*$/;

  return (value: string): ValidationResult<string> => {
    if (typeof value !== 'string') {
      return failure('Value must be a string');
    }

    if (value === '') {
      return failure('Value cannot be empty');
    }

    if (!slugRegex.test(value)) {
      return failure(errorMessage);
    }

    return success(value);
  };
}

export interface IsInOptions {
  caseSensitive?: boolean;
  errorMessage?: string;
}

/**
 * Validates that a value is in an allowed set.
 */
export function isIn(allowedOptions: readonly string[], options: IsInOptions = {}): Validator<string, string> {
  const { caseSensitive = true } = options;

  const optionSet = new Set(
    caseSensitive ? allowedOptions : allowedOptions.map(o => o.toLowerCase())
  );

  return (value: string): ValidationResult<string> => {
    if (typeof value !== 'string') {
      return failure('Value must be a string');
    }

    const checkValue = caseSensitive ? value : value.toLowerCase();

    if (!optionSet.has(checkValue)) {
      return failure(options.errorMessage ?? `Value must be one of: ${allowedOptions.join(', ')}`);
    }

    return success(value);
  };
}

export interface TrimmedOptions {
  allowEmpty?: boolean;
}

/**
 * Validates and trims whitespace from a string.
 */
export function trimmed(options: TrimmedOptions = {}): Validator<string, string> {
  const { allowEmpty = false } = options;

  return (value: string): ValidationResult<string> => {
    if (typeof value !== 'string') {
      return failure('Value must be a string');
    }

    const result = value.trim();

    if (!allowEmpty && result === '') {
      return failure('Value cannot be empty');
    }

    return success(result);
  };
}

File: shared/shared/go_libs/validation/string.go
package validation

import (
	"regexp"
	"strings"
	"unicode"
)

// StringValidator is a validator specifically for string values.
type StringValidator struct {
	validate func(string) ValidationResult
}

// Validate implements the Validator interface.
func (v *StringValidator) Validate(value any) ValidationResult {
	s, ok := value.(string)
	if !ok {
		return Failure("value must be a string")
	}
	return v.validate(s)
}

// NotEmpty validates that a string is not empty or whitespace-only.
// The returned value is trimmed.
func NotEmpty(opts ...func(*notEmptyOpts)) Validator {
	o := &notEmptyOpts{
		errorMessage: "value cannot be empty",
	}
	for _, opt := range opts {
		opt(o)
	}

	return &StringValidator{
		validate: func(s string) ValidationResult {
			trimmed := strings.TrimSpace(s)
			if trimmed == "" {
				return Failure(o.errorMessage)
			}
			return Success(trimmed)
		},
	}
}

type notEmptyOpts struct {
	errorMessage string
}

// WithNotEmptyError sets a custom error message for NotEmpty.
func WithNotEmptyError(msg string) func(*notEmptyOpts) {
	return func(o *notEmptyOpts) {
		o.errorMessage = msg
	}
}

// Length validates that a string length is within a range.
func Length(minLen, maxLen int, opts ...func(*lengthOpts)) Validator {
	o := &lengthOpts{}
	for _, opt := range opts {
		opt(o)
	}

	return &StringValidator{
		validate: func(s string) ValidationResult {
			l := len(s)
			if l < minLen {
				if o.errorMessage != "" {
					return Failure(o.errorMessage)
				}
				return Failuref("value must be at least %d characters", minLen)
			}
			if maxLen > 0 && l > maxLen {
				if o.errorMessage != "" {
					return Failure(o.errorMessage)
				}
				return Failuref("value must be at most %d characters", maxLen)
			}
			return Success(s)
		},
	}
}

type lengthOpts struct {
	errorMessage string
}

// WithLengthError sets a custom error message for Length.
func WithLengthError(msg string) func(*lengthOpts) {
	return func(o *lengthOpts) {
		o.errorMessage = msg
	}
}

// MinLength validates minimum string length.
func MinLength(min int, opts ...func(*lengthOpts)) Validator {
	return Length(min, 0, opts...)
}

// MaxLength validates maximum string length.
func MaxLength(max int, opts ...func(*lengthOpts)) Validator {
	return Length(0, max, opts...)
}

// Match validates that a string matches a regex pattern.
func Match(pattern string, opts ...func(*matchOpts)) Validator {
	o := &matchOpts{
		errorMessage: "value does not match required pattern",
	}
	for _, opt := range opts {
		opt(o)
	}

	re := regexp.MustCompile(pattern)

	return &StringValidator{
		validate: func(s string) ValidationResult {
			if !re.MatchString(s) {
				return Failure(o.errorMessage)
			}
			return Success(s)
		},
	}
}

type matchOpts struct {
	errorMessage string
}

// WithMatchError sets a custom error message for Match.
func WithMatchError(msg string) func(*matchOpts) {
	return func(o *matchOpts) {
		o.errorMessage = msg
	}
}

// Alphanumeric validates that a string contains only alphanumeric characters.
func Alphanumeric(opts ...func(*alphanumericOpts)) Validator {
	o := &alphanumericOpts{
		allowUnderscore: false,
		allowDash:       false,
		errorMessage:    "value must contain only alphanumeric characters",
	}
	for _, opt := range opts {
		opt(o)
	}

	return &StringValidator{
		validate: func(s string) ValidationResult {
			if s == "" {
				return Failure("value cannot be empty")
			}

			for _, r := range s {
				if unicode.IsLetter(r) || unicode.IsDigit(r) {
					continue
				}
				if o.allowUnderscore && r == '_' {
					continue
				}
				if o.allowDash && r == '-' {
					continue
				}
				return Failure(o.errorMessage)
			}
			return Success(s)
		},
	}
}

type alphanumericOpts struct {
	allowUnderscore bool
	allowDash       bool
	errorMessage    string
}

// WithUnderscore allows underscores in Alphanumeric.
func WithUnderscore() func(*alphanumericOpts) {
	return func(o *alphanumericOpts) {
		o.allowUnderscore = true
	}
}

// WithDash allows dashes in Alphanumeric.
func WithDash() func(*alphanumericOpts) {
	return func(o *alphanumericOpts) {
		o.allowDash = true
	}
}

// WithAlphanumericError sets a custom error message for Alphanumeric.
func WithAlphanumericError(msg string) func(*alphanumericOpts) {
	return func(o *alphanumericOpts) {
		o.errorMessage = msg
	}
}

// Slug validates that a string is a valid URL slug.
// A valid slug contains only lowercase letters, numbers, and hyphens,
// starts and ends with alphanumeric, and has no consecutive hyphens.
func Slug(opts ...func(*slugOpts)) Validator {
	o := &slugOpts{
		errorMessage: "value must be a valid URL slug",
	}
	for _, opt := range opts {
		opt(o)
	}

	slugRegex := regexp.MustCompile(`^[a-z0-9]+(?:-[a-z0-9]+)*$`)

	return &StringValidator{
		validate: func(s string) ValidationResult {
			if s == "" {
				return Failure("value cannot be empty")
			}
			if !slugRegex.MatchString(s) {
				return Failure(o.errorMessage)
			}
			return Success(s)
		},
	}
}

type slugOpts struct {
	errorMessage string
}

// WithSlugError sets a custom error message for Slug.
func WithSlugError(msg string) func(*slugOpts) {
	return func(o *slugOpts) {
		o.errorMessage = msg
	}
}

// In validates that a value is in an allowed set.
func In(options []string, opts ...func(*inOpts)) Validator {
	o := &inOpts{
		caseSensitive: true,
	}
	for _, opt := range opts {
		opt(o)
	}

	// Build set for fast lookup
	optionSet := make(map[string]struct{}, len(options))
	for _, opt := range options {
		key := opt
		if !o.caseSensitive {
			key = strings.ToLower(opt)
		}
		optionSet[key] = struct{}{}
	}

	return &StringValidator{
		validate: func(s string) ValidationResult {
			checkValue := s
			if !o.caseSensitive {
				checkValue = strings.ToLower(s)
			}

			if _, ok := optionSet[checkValue]; !ok {
				if o.errorMessage != "" {
					return Failure(o.errorMessage)
				}
				return Failuref("value must be one of: %s", strings.Join(options, ", "))
			}
			return Success(s)
		},
	}
}

type inOpts struct {
	caseSensitive bool
	errorMessage  string
}

// CaseInsensitive makes In comparison case-insensitive.
func CaseInsensitive() func(*inOpts) {
	return func(o *inOpts) {
		o.caseSensitive = false
	}
}

// WithInError sets a custom error message for In.
func WithInError(msg string) func(*inOpts) {
	return func(o *inOpts) {
		o.errorMessage = msg
	}
}

// Trimmed validates and trims whitespace from a string.
func Trimmed(opts ...func(*trimmedOpts)) Validator {
	o := &trimmedOpts{
		allowEmpty: false,
	}
	for _, opt := range opts {
		opt(o)
	}

	return &StringValidator{
		validate: func(s string) ValidationResult {
			trimmed := strings.TrimSpace(s)
			if !o.allowEmpty && trimmed == "" {
				return Failure("value cannot be empty")
			}
			return Success(trimmed)
		},
	}
}

type trimmedOpts struct {
	allowEmpty bool
}

// AllowEmpty allows empty strings in Trimmed.
func AllowEmpty() func(*trimmedOpts) {
	return func(o *trimmedOpts) {
		o.allowEmpty = true
	}
}
